src/online2/online-nnet3-wake-word-faster-decoder.h

```diff
@@ -0,0 +1,145 @@
+// the _maximum_ beam value allowed.
+struct OnlineWakeWordFasterDecoderOpts : public FasterDecoderOptions {
+  BaseFloat rt_min; // minimum decoding runtime factor
+  BaseFloat rt_max; // maximum decoding runtime factor
+  int32 batch_size; // number of features decoded in one go
+  int32 inter_utt_sil; // minimum silence (#frames) to trigger end of utterance
+  int32 max_utt_len_; // if utt. is longer, we accept shorter silence as utt. separators
+  int32 update_interval; // beam update period in # of frames
+  BaseFloat beam_update; // rate of adjustment of the beam
+  BaseFloat max_beam_update; // maximum rate of beam adjustment
+
+  OnlineWakeWordFasterDecoderOpts() :
+    rt_min(.7), rt_max(.75), batch_size(27),
+    inter_utt_sil(50), max_utt_len_(1500),
+    update_interval(3), beam_update(.01),
+    max_beam_update(0.05) {}
+
+  void Register(OptionsItf *opts, bool full) {
+    FasterDecoderOptions::Register(opts, full);
+    opts->Register("rt-min", &rt_min,
+                   "Approximate minimum decoding run time factor");
+    opts->Register("rt-max", &rt_max,
+                   "Approximate maximum decoding run time factor");
+    opts->Register("update-interval", &update_interval,
+                   "Beam update interval in frames");
+    opts->Register("beam-update", &beam_update, "Beam update rate");
+    opts->Register("max-beam-update", &max_beam_update, "Max beam update rate");
+    opts->Register("inter-utt-sil", &inter_utt_sil,
+                   "Maximum # of silence frames to trigger new utterance");
+    opts->Register("max-utt-length", &max_utt_len_,
+                   "If the utterance becomes longer than this number of frames, "
+                   "shorter silence is acceptable as an utterance separator");
+  }
+};
+
+class OnlineWakeWordFasterDecoder : public FasterDecoder {
+ public:
+  // Codes returned by Decode() to show the current state of the decoder
+  enum DecodeState {
+    kEndFeats = 1, // No more scores are available from the Decodable
+    kEndUtt = 2, // End of utterance, caused by e.g. a sufficiently long silence
+    kEndBatch = 4 // End of batch - end of utterance not reached yet
+  };
+
+  // "sil_phones" - the IDs of all silence phones
+  OnlineWakeWordFasterDecoder(const fst::Fst<fst::StdArc> &fst,
+                              const OnlineWakeWordFasterDecoderOpts &opts,
+                              const std::vector<int32> &sil_phones,
+                              const TransitionModel &trans_model,
+                              const nnet3::DecodableNnetSimpleLoopedInfo &info,
+                              OnlineNnet2FeaturePipeline *features)
+      : FasterDecoder(fst, opts), opts_(opts),
+        silence_set_(sil_phones), trans_model_(trans_model),
+        max_beam_(opts.beam), effective_beam_(FasterDecoder::config_.beam),
+        state_(kEndFeats), frame_(0), utt_frames_(0),
+        input_feature_frame_shift_in_seconds_(features->FrameShiftInSeconds()),
```

Comment:
I'm hoping the features wouldn't need to be given to the constructor.   This should probably be simplified, and only the decodable object given to the constructor; the decodable can be constructed outside of here.
The basic decoder would be very similar to lattice-faster-decoder in its interface, except it would have the immortal-token stuff and maybe keep track of the word sequence we've seen.  It might be easier to try to understand the algorithm and copy/modify LatticeFasterDecoder.  Not sure.